### PR TITLE
refactor(return code): remove os.Exit from library

### PR DESCRIPTION
**Pull request by tim-shockley**

### CODE REVIEW

File: pkg/utils/constants.go [hunk 1]
```diff
@@ -19,5 +19,7 @@ const (
 	LMSPort = "16992"
 
 	// MPSServerMaxLength is the max length of the servername
-	MPSServerMaxLength = 256
+	MPSServerMaxLength    = 256
+	ReturnCode_SUCCESS    = 0
```
Comment (rsdmike) [inline pkg/utils/constants.go:23]: Nice that these are constants. Can you update the names to be just SUCCESS or FAIL, in any case -- remove the underscores and use mixedCase.

**Review by rsdmike — CHANGES_REQUESTED** [review]
Are you working on unit tests updates?

File: README.md [hunk 2]
```diff
@@ -28,7 +28,7 @@ go build -o rpc ./cmd/main.go
 
 #### As Library: 
 ```
-go build -buildmode=c-shared -o rpc.so ./cmd   
+go build -buildmode=c-shared -o librpc.so.1.0 ./cmd
```
Comment (matt-primrose) [inline README.md:31]: Why are we renaming this?

File: cmd/main.go [hunk 3]
```diff
@@ -4,32 +4,55 @@
...
 	"rpc/internal/client"
 	"rpc/internal/rps"
+	"rpc/pkg/utils"
+	"strings"
 
 	log "github.com/sirupsen/logrus"
 )
 
-func runRPC(args []string) {
+func checkAccess() (int, error) {
```
Comment (matt-primrose) [inline cmd/main.go:20]: Why are we moving everything from lib.go into main.go?  What was wrong with it being the way it was before?  Seems out of scope for the user story

File: internal/lm/engine.go [hunk 4]
```diff
@@ -40,7 +40,7 @@ func NewLMEConnection(data chan []byte, errors chan error, status chan bool) *LM
 func (lme *LMEConnection) Initialize() error {
 	err := lme.Command.Open(true)
 	if err != nil {
-		log.Fatal(err)
+		log.Error(err)
```
Comment (matt-primrose) [inline internal/lm/engine.go:43]: need unit tests for this

File: internal/lm/engine.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ func (lme *LMEConnection) Initialize() error {
 
 	err = lme.execute(bin_buf)
 	if err != nil {
-		log.Fatal(err)
+		log.Error(err)
```
Comment (matt-primrose) [inline internal/lm/engine.go:53]: need unit tests for this

File: internal/lm/service.go [hunk 6]
```diff
@@ -33,8 +34,7 @@ func NewLMSConnection(address string, port string, data chan []byte, errors chan
 	return lms
 }
 func (lms *LMSConnection) Initialize() error {
-	log.Fatal("Not Implemented")
-	return nil
+	return errors.New("not implemented")
```
Comment (matt-primrose) [inline internal/lm/service.go:37]: need unit tests for this

File: internal/rps/rps.go [hunk 7]
```diff
@@ -27,15 +27,9 @@ func NewAMTActivationServer(url string) AMTActivationServer {
 	return amtactivationserver
 }
 
-func PrepareInitialMessage(flags *rpc.Flags) Message {
-	//create activation request
+func PrepareInitialMessage(flags *rpc.Flags) (Message, error) {
```
Comment (matt-primrose) [inline internal/rps/rps.go:30]: need unit tests for this

File: internal/rps/rps.go [hunk 8]
```diff
@@ -27,15 +27,9 @@ func NewAMTActivationServer(url string) AMTActivationServer {
...
 	payload := NewPayload()
-
-	messageRequest, err := payload.CreateMessageRequest(*flags)
-	if err != nil {
-		log.Fatal(err)
-	}
-	return messageRequest
+	return payload.CreateMessageRequest(*flags)
```
Comment (matt-primrose) [inline internal/rps/rps.go:32]: need unit tests for this

File: README.md (hunk 2 above)
Comment (tim-shockley) [reply README.md:31]: to follow linux standard .so library format. is there some existing Linux usage that is not documented?

File: cmd/lib.go [hunk 9]
```diff
@@ -2,28 +2,37 @@ package main
...
-func checkAccess() {
-	amt := amt.NewAMTCommand()
-	result, err := amt.Initialize()
+func checkAccess() int {
+	amtCommand := amt.NewAMTCommand()
+	result, err := amtCommand.Initialize()
 	if !result || err != nil {
-		println("Unable to launch application. Please ensure that Intel ME is present, the MEI driver is installed and that this application is run with administrator or root privileges.")
-		os.Exit(1)
+		log.Error("Unable to launch application. " +
```
[the remark this answers is not in the document]
Comment (tim-shockley) [reply cmd/lib.go]: Readability, long lines.

File: cmd/main.go (hunk 3 above)
Comment (tim-shockley) [reply cmd/main.go:20]: methods were redundant duplication. why was there a cmd/lib.go to create a user library in the cmd namespace in the first place. which calls back into the main.go file anyways? and you don't 'execute' a library, you link to it. so not sure how that ended up in the cmd/ folder to start with?

File: internal/lm/engine.go (hunk 4 above)
Comment (tim-shockley) [reply internal/lm/engine.go:43]: weren't there originally when I started on this story but I'll add them. not sure why they weren't there first?

File: cmd/lib.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rsdmike) [reply cmd/lib.go]: Please use backticks instead of string concatenation

File: cmd/main.go (hunk 3 above)
Comment (rsdmike) [reply cmd/main.go:20]: The reason is for avoiding gcc requirement on either OS if not using as a library. Please put it back.

File: README.md (hunk 2 above)
Comment (tim-shockley) [reply README.md]: I ended up taking off the version ID part so the example is less brittle. On a standard linux distribution and install, the .so get a lot of links created in the common library file with version extenstions. That's way more of an end user situation than the sample needs actually, but for me coming from Linux, I was expecting to see a librpc.so file somewhere in the project. Thanks.

File: cmd/main.go (hunk 3 above)
Comment (tim-shockley) [reply cmd/main.go:22]: done, tested executable and sample on Ubuntu, 
tested  only executable on Win ... no gcc installed yet

File: internal/lm/engine.go (hunk 4 above)
Comment (tim-shockley) [reply internal/lm/engine.go:43]: done

File: internal/lm/engine.go (hunk 5 above)
Comment (tim-shockley) [reply internal/lm/engine.go:53]: done

File: internal/lm/service.go (hunk 6 above)
Comment (tim-shockley) [reply internal/lm/service.go:37]: done

File: internal/rps/rps.go (hunk 7 above)
Comment (tim-shockley) [reply internal/rps/rps.go:30]: I can't figure out a way to test this as written, the calls in that function are covered under message_test.go, would need to add an interface capability probably? could move it into the executor code .makeItSo() but that has no test coverage currently either.

File: internal/rps/rps.go (hunk 8 above)
Comment (tim-shockley) [reply internal/rps/rps.go:32]: same comment as previous.

File: pkg/utils/constants.go (hunk 1 above)
Comment (tim-shockley) [reply pkg/utils/constants.go]: done

File: cmd/lib.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tim-shockley) [reply cmd/lib.go]: should clarify, readability in the source only, I wasn't intending to add linebreaks into the error message. Is that what you are suggesting?

File: cmd/main.go [hunk 10]
```diff
@@ -54,20 +79,21 @@ func handleFlags(args []string) *rpc.Flags {
...
+			log.Error(err.Error())
+		}
+		log.Error(AccessErrMsg)
+		return
+	}
+	status, err = runRPC(os.Args)
+	if err != nil {
+		log.Error(err.Error())
 	}
+	log.Info("Completed with status code: ", status)
```
Comment (rsdmike) [inline cmd/main.go:98]: I think this needs to remain os.Exit(status) -- this call must happen otherwise 0 is always the return code regardless of what we log it to be. It is only the runRPC function that is used by the library so what happens in main here can remain as os.Exit. I think we also should remove the log statement it doesn't tell us anything different than what the OS would tell me if you use os.exit as we should. If you want to keep it, put it as trace level. But if you do decide to keep it, you'll need to inspect the code and accurately log an error if it is not 0 or trace if it is 0

**Review by rsdmike — CHANGES_REQUESTED** [review]

File: cmd/main.go (hunk 10 above)
Comment (tim-shockley) [reply cmd/main.go:98]: ok, makes sense, I'll fix it.

**Review by rsdmike — APPROVED** [review]

**Review by matt-primrose — APPROVED** [review]
